File: backend/scripts/import_contacts.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const csvPath = process.argv[2];
if (!csvPath) {
  console.error('Usage: node import_contacts.js <contacts.csv>');
  process.exit(2);
}

const root = path.resolve(__dirname, '..');
const contactsJson = path.join(root, 'data', 'contacts.json');
const logFile = path.join(root, 'import_debug_node.log');

function dlog(msg) {
  try { fs.appendFileSync(logFile, msg + '\n', 'utf8'); } catch (e) {}
}

function normalizePhone(s) {
  if (!s) return '';
  let digits = s.replace(/\D/g, '');
  if (!digits) return '';
  if (digits.length === 10) return '91' + digits;
  if (digits.length === 12 && digits.startsWith('91')) return digits;
  if (digits.length === 11 && digits.startsWith('0')) return '91' + digits.slice(1);
  return digits;
}

async function run() {
  dlog(`START node import: csv=${csvPath} contacts_json=${contactsJson}`);
  let existing = [];
  try {
    if (fs.existsSync(contactsJson)) {
      existing = JSON.parse(fs.readFileSync(contactsJson, 'utf8')) || [];
    }
  } catch (e) {
    dlog('Error reading existing contacts: ' + e.message);
    existing = [];
  }

  const existingMobiles = new Set(existing.map(c => normalizePhone(c.mobile || c.phone || '')));
  const added = [];
  let rows = 0;

  const rl = readline.createInterface({ input: fs.createReadStream(csvPath, { encoding: 'utf8' }) });
  for await (const line of rl) {
    rows++;
    // try to find phone-like substrings
    const phones = [];
    const found = line.match(/\d[\d\s\-:\+]{6,}\d/g);
    if (found) {
      for (const f of found) {
        const p = normalizePhone(f);
        if (p) phones.push(p);
      }
    }
    if (phones.length === 0) continue;
    const primary = phones[0];
    if (!primary) continue;
    if (existingMobiles.has(primary)) continue;
    // name: take leading text before first comma
    const name = line.split(',')[0].replace(/\"/g, '').trim() || primary;
    const contact = {
      id: Date.now() + rows,
      name,
      mobile: primary,
      company: '',
      city: '',
      state: '',
      vehicle_type: '',
      is_active: 1,
      created_at: new Date().toISOString()
    };
    added.push(contact);
    existingMobiles.add(primary);
  }

  dlog(`Rows scanned: ${rows}, Added: ${added.length}`);
  if (added.length === 0) {
    console.log('No new contacts found to import.');
    return;
  }

  const newList = added.concat(existing);
  try {
    fs.mkdirSync(path.dirname(contactsJson), { recursive: true });
    fs.writeFileSync(contactsJson, JSON.stringify(newList, null, 2), 'utf8');
    dlog(`Imported ${added.length} contacts`);
    console.log(`Imported ${added.length} contacts`);
  } catch (e) {
    dlog('Error writing contacts.json: ' + e.message);
    console.error('Failed to write contacts.json:', e.message);
    process.exit(1);
  }
}

run().catch(err => {
  dlog('Unhandled error: ' + err.message);
  console.error(err);
  process.exit(1);
});

File: import_contacts.py
#!/usr/bin/env python3
import csv
import json
import os
import re
import sys
from datetime import datetime


def normalize_phone(value):
    s = re.sub(r"\D", "", str(value or ""))
    if not s:
        return ""
    if s.startswith('91') and len(s) == 12:
        return s
    if len(s) == 10:
        return '91' + s
    if len(s) == 11 and s.startswith('0'):
        return '91' + s[1:]
    return s


def extract_phones_from_row(row):
    phones = []
    # look for columns with phone in header
    for k, v in row.items():
        if v is None:
            continue
        if re.search(r'phone|mobile|contact|wa|whats', k, re.IGNORECASE):
            p = normalize_phone(v)
            if p:
                phones.append(p)

    # fallback: search any cell for digit sequences
    if not phones:
        for v in row.values():
            if not v:
                continue
            found = re.findall(r"\d{7,13}", str(v))
            for f in found:
                p = normalize_phone(f)
                if p:
                    phones.append(p)

    # dedupe
    seen = []
    for p in phones:
        if p not in seen:
            seen.append(p)
    return seen


def build_name(row):
    parts = []
    for k in ['First Name', 'Middle Name', 'Last Name', 'FirstName', 'LastName', 'Name', 'name', 'Full Name', 'FullName']:
        if k in row and row[k]:
            parts.append(str(row[k]).strip())
    if parts:
        return ' '.join(parts)
    # fallback: use any 'name' like column
    for k, v in row.items():
        if k and 'name' in k.lower() and v:
            return str(v).strip()
    return ''


def main():
    if len(sys.argv) < 2:
        print('Usage: import_contacts.py <contacts.csv>')
        sys.exit(2)

    csv_path = sys.argv[1]
    root = os.path.dirname(os.path.abspath(__file__))
    contacts_json = os.path.join(root, 'backend', 'data', 'contacts.json')
    debug_log = os.path.join(root, 'import_debug.log')

    def dlog(msg):
        try:
            with open(debug_log, 'a', encoding='utf-8') as dl:
                dl.write(msg + "\n")
        except Exception:
            pass

    dlog(f"START import: csv={csv_path} contacts_json={contacts_json}")

    if not os.path.exists(csv_path):
        dlog(f'CSV not found: {csv_path}')
        print('CSV not found:', csv_path)
        sys.exit(1)

    # load existing
    if os.path.exists(contacts_json):
        try:
            with open(contacts_json, 'r', encoding='utf-8') as f:
                existing = json.load(f)
        except Exception:
            existing = []
    else:
        existing = []

    existing_mobiles = set([normalize_phone(c.get('mobile') or c.get('phone') or '') for c in existing])

    added = []
    skipped = 0
    rows_seen = 0

    with open(csv_path, 'r', encoding='utf-8', errors='replace') as fh:
        reader = csv.DictReader(fh)
        for idx, row in enumerate(reader):
            rows_seen += 1
            name = build_name(row)
            phones = extract_phones_from_row(row)
            if not phones:
                continue
            primary = phones[0]
            norm = normalize_phone(primary)
            if not norm:
                continue
            if norm in existing_mobiles:
                skipped += 1
                continue

            contact = {
                'id': int(datetime.utcnow().timestamp() * 1000) + idx,
                'name': name or norm,
                'mobile': norm,
                'company': '',
                'city': '',
                'state': '',
                'vehicle_type': '',
                'is_active': 1,
                'created_at': datetime.utcnow().isoformat() + 'Z'
            }
            added.append(contact)
            existing_mobiles.add(norm)

    dlog(f'Rows scanned: {rows_seen}, Added: {len(added)}, Skipped duplicates: {skipped}')

    if not added:
        dlog('No new contacts found to import.')
        print(f'No new contacts found to import. Skipped duplicates: {skipped}')
        return

    # prepend new contacts
    new_list = added + existing
    os.makedirs(os.path.dirname(contacts_json), exist_ok=True)
    with open(contacts_json, 'w', encoding='utf-8') as f:
        json.dump(new_list, f, indent=2, ensure_ascii=False)
    dlog(f'Imported {len(added)} contacts, skipped {skipped} duplicates. Written to {contacts_json}')
    print(f'Imported {len(added)} contacts, skipped {skipped} duplicates. Written to {contacts_json}')


if __name__ == '__main__':
    main()
